nni/retiarii/oneshot/pytorch/utils.py

```diff
@@ -180,3 +180,120 @@ def replace_input_choice(root_module, init_fn, modules=None):
         A list from layer choice keys (names) and replaced modules.
     """
     return _replace_module_with_type(root_module, init_fn, (InputChoice, nn.InputChoice), modules)
+
+
+def replace_module_with_type(root_module, init_fn, type_name, modules):
+
+    return _replace_module_with_type(root_module, init_fn, type_name, modules)
+
+
+class ParallelTrainValDataLoader(DataLoader):
+    """
+    A dataloader yields both train data and validation data in a batch, with an order of
+    (train_batch, val_batch). The shorter one will be upsampled (repeated) to the length
+    of the longer one, and the tail of the last repeat will be dropped.
+
+    Some NAS algorithms, i.e. DARTS and Proxyless, require this type of dataloader.
+
+    Parameters
+    ----------
+    train_data : DataLoader
+        training dataloader
+    val_data : DataLoader
+        validation dataloader
+
+    Example
+    --------
+    Fit your dataloaders into a parallel one.
+    >>> para_loader = ParallelTrainValDataLoader(train_dataloader, val_datalodaer)
+    Then you can use the 'para_loader' as a normal training loader.
+    """
+    def __init__(self, train_dataloader, val_dataloader):
+        self.train_loader = train_dataloader
+        self.val_loader = val_dataloader
+        self.equal_len = len(train_dataloader) == len(val_dataloader)
+        self.train_longger = len(train_dataloader) > len(val_dataloader)
+        super().__init__(None)
+
+    def __iter__(self):
+        self.train_iter = iter(self.train_loader)
+        self.val_iter = iter(self.val_loader)
+        return self
+
+    def __next__(self):
+        try:
+            train_batch = next(self.train_iter)
+        except StopIteration:
+            # training data is used up
+            if self.equal_len or self.train_longger:
+                # if training is the longger one or equal, stop iteration
+                raise StopIteration()
+            # if training is the shorter one, upsample it
+            self.train_iter = iter(self.train_loader)
+            train_batch = next(self.train_iter)
+
+        try:
+            val_batch = next(self.val_iter)
+        except StopIteration:
+            # validation data is used up
+            if not self.train_longger:
+                # if validation is the longger one (the equal condition is
+                # covered above), stop iteration
+                raise StopIteration()
+            # if validation is the shorter one, upsample it
+            self.val_iter = iter(self.val_loader)
+            val_batch = next(self.val_iter)
+
+        return train_batch, val_batch
+
+    def __len__(self) -> int:
+        return max(len(self.train_loader), len(self.val_loader))
+
+
+class ConcatenateTrainValDataLoader(DataLoader):
+    """
+    A dataloader that yields validation data after training data in an epoch. You will
+    got a batch in a form of (batch, b) in the training step, where the bool variable 'b'
```

Comment:
Suggest `(batch, source)` where `source` can be `train` and `val`.
This is because we might have `test` in future.